ratis-server/src/main/java/org/apache/ratis/statemachine/impl/TransactionContextImpl.java

```diff
@@ -209,4 +214,8 @@ public TransactionContext cancelTransaction() throws IOException {
     // call this to let the SM know that Transaction cannot be synced
     return stateMachine.cancelTransaction(this);
   }
+
+  public static LogEntryProto getLogEntry(TransactionContext context) {
+    return ((TransactionContextImpl) context).logEntry;
+  }
```

Comment:
Let's move this to `TransactionContext`
```java
//TransactionContext.java
  /**
   * @return the committed log entry if it exists; otherwise, returns null.
   *         The returned value is safe to use only before {@link StateMachine#applyTransaction} returns.
   *         Once {@link StateMachine#applyTransaction} has returned, it is unsafe to use the log entry
   *         since the underlying buffers can possiby be released.
   */
  default LogEntryProto getLogEntryUnsafe() {
    return getLogEntryRef().get();
  }
```